File: LeetCode/Strings/28-Find-the-Index-of-the-First-Occurrence-in-a-String.py
"""
Problem: 28. Find the Index of the First Occurrence in a String
Platform: LeetCode
Difficulty: Easy

Approach:
Brute Force String Matching

Try every possible starting position in the haystack.
Compare the substring with the needle.
Return the first matching index.

Time Complexity: O((n - m + 1) × m)
Space Complexity: O(1)
"""

class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        n = len(haystack)
        m = len(needle)

        for i in range(n - m + 1):
            if haystack[i:i + m] == needle:
                return i

        return -1

File: LeetCode/Strings/28-Find-the-Index-of-the-First-Occurrence-in-a-String.js
/**
 * Problem: 28. Find the Index of the First Occurrence in a String
 * Platform: LeetCode
 * Difficulty: Easy
 *
 * Approach:
 * Brute Force String Matching
 *
 * Check every possible starting index in the haystack.
 * If every character of the needle matches,
 * return the current index.
 *
 * Time Complexity: O((n - m + 1) × m)
 * Space Complexity: O(1)
 */

/**
 * @param {string} haystack
 * @param {string} needle
 * @return {number}
 */
var strStr = function(haystack, needle) {
    const n = haystack.length;
    const m = needle.length;

    for (let i = 0; i <= n - m; i++) {
        let j = 0;

        while (j < m && haystack[i + j] === needle[j]) {
            j++;
        }

        if (j === m) {
            return i;
        }
    }

    return -1;
};